Realm/ObjectServerTests/RLMFlexibleSyncServerTests.mm

```diff
@@ -0,0 +1,1014 @@
+    XCTAssertEqualObjects(foundSubscription2.objectClassName, @"Person");
+}
+
+- (void)testUpdateSubscriptionQueryWithDifferentClassName {
+    RLMRealm *realm = [self openFlexibleSyncRealm:_cmd];
+    RLMSyncSubscriptionSet *subs = realm.subscriptions;
+
+    [subs write:^{
+        [subs addSubscriptionWithClassName:Person.className
+                          subscriptionName:@"subscription_1"
+                                     where:@"age > 15"];
+    }];
+
+    XCTAssertEqual(subs.version, 1UL);
+    XCTAssertEqual(subs.count, 1UL);
+
+    RLMSyncSubscription *foundSubscription = [subs subscriptionWithName:@"subscription_1"];
+    XCTAssertNotNil(foundSubscription);
+
+    [subs write:^{
+        [foundSubscription updateSubscriptionWithClassName:Dog.className predicate:[NSPredicate predicateWithFormat:@"name == 'Tomas'"]];
+    }];
+
+    XCTAssertEqual(subs.version, 2UL);
+    XCTAssertEqual(subs.count, 1UL);
+
+    RLMSyncSubscription *foundSubscription2 = [subs subscriptionWithName:@"subscription_1"];
+    XCTAssertNotNil(foundSubscription2);
+    XCTAssertEqualObjects(foundSubscription2.queryString, @"name == \"Tomas\"");
+    XCTAssertEqualObjects(foundSubscription2.objectClassName, @"Dog");
+}
+
+- (void)testUpdateSubscriptionQueryWithoutWriteThrow {
+    RLMRealm *realm = [self openFlexibleSyncRealm:_cmd];
+    RLMSyncSubscriptionSet *subs = realm.subscriptions;
+
+    [subs write:^{
+        [subs addSubscriptionWithClassName:Person.className
+                          subscriptionName:@"subscription_1"
+                                     where:@"age > 15"];
+    }];
+
+    XCTAssertEqual(subs.version, 1UL);
+    XCTAssertEqual(subs.count, 1UL);
+
+    RLMSyncSubscription *foundSubscription = [subs subscriptionWithName:@"subscription_1"];
+    XCTAssertNotNil(foundSubscription);
+
+    RLMAssertThrowsWithReason([foundSubscription updateSubscriptionWithClassName:Dog.className predicate:[NSPredicate predicateWithFormat:@"name == 'Tomas'"]], @"Can only add, remove, or update subscriptions within a write subscription block.");
+}
+
+- (void)testSubscriptionSetIterate {
+    RLMRealm *realm = [self openFlexibleSyncRealm:_cmd];
+    RLMSyncSubscriptionSet *subs = realm.subscriptions;
+
+    double numberOfSubs = 100;
+    [subs write:^{
+        for (int i = 0; i < numberOfSubs; ++i) {
+            [subs addSubscriptionWithClassName:Person.className
+                              subscriptionName:[NSString stringWithFormat:@"person_age_%d", i]
+                                         where:[NSString stringWithFormat:@"age > %d", i]];
+        }
+    }];
+
+    XCTAssertEqual(subs.version, 1UL);
+    XCTAssertEqual(subs.count, (unsigned long)numberOfSubs);
+
+    __weak id objects[(unsigned long)pow(numberOfSubs, 2.0) + (unsigned long)numberOfSubs];
+    NSInteger count = 0;
+    for (RLMSyncSubscription *sub in subs) {
+        XCTAssertNotNil(sub);
+        objects[count++] = sub;
+        for (RLMSyncSubscription *sub in subs) {
```

Comment:
The nested for-loop tests that we implemented NSFastEnumeration correctly. It was previously implemented incorrectly and a nested for-loop would have broken it. The use of `pow` was to dynamically calculate count since the loop is O(n²).